pkg/kubectl/cmd/util/discovery.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// lazyObject defers loading the mapper and typer until necessary.
+type lazyObject struct {
+	lock   sync.Mutex
+	loader func() (meta.RESTMapper, runtime.ObjectTyper, error)
+
+	err    error
+	mapper meta.RESTMapper
+	typer  runtime.ObjectTyper
+}
+
+// NewLazyObjectLoader handles unrecoverable errors when creating a RESTMapper / ObjectTyper by
+// returning those initialization errors when the interface methods are invoked. This defers the
+// initialization and any server calls until a client actually needs to perform the action.
+func NewLazyObjectLoader(fn func() (meta.RESTMapper, runtime.ObjectTyper, error)) (meta.RESTMapper, runtime.ObjectTyper) {
+	obj := &lazyObject{loader: fn}
+	return obj, obj
+}
+
+// init lazily loads the mapper and typer, returning an error if initialization has failed.
+func (o *lazyObject) init() error {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.err != nil || o.loader == nil {
+		return o.err
+	}
+	o.mapper, o.typer, o.err = o.loader()
+	o.loader = nil
```

Comment:
I don't see a need to do this.  mapper and typer shouldn't be nil out of the loader.

Reply:
This is the indicator that covers whether loading has happened.

Reply:
> This is the indicator that covers whether loading has happened.

Using it as the trigger opens you up to accidentally leaving nil.  Seems like triggering the other way around avoids that problem.